### PR TITLE
[CodeCamp2023-472] Add parameter save_begin

### PR DESCRIPTION
**Pull request by KerwinKai**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add parameter save_begin to control the epoch number or iteration number at which checkpoint saving begins. As issue https://github.com/open-mmlab/mmengine/issues/1192 descriped.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: tests/test_hooks/test_checkpoint_hook.py [hunk 1]
```diff
@@ -632,3 +632,23 @@ def test_with_runner(self, training_type):
...
+        cfg = copy.deepcopy(common_cfg)
+        cfg.default_hooks.checkpoint.interval = 2
+        cfg.default_hooks.checkpoint.save_begin = 6
+        runner = self.build_runner(cfg)
+        runner.train()
+        for i in range(6):
+            self.assertFalse(
+                osp.isfile(osp.join(cfg.work_dir, f'{training_type}_{i}.pth')))
+        for i in range(6, 11):
+            if i % 2:
```
Comment (HAOCHENYE) [inline tests/test_hooks/test_checkpoint_hook.py:646]: Refer to https://google.github.io/styleguide/pyguide.html#214-truefalse-evaluations
```suggestion
            if i % 2 == 1:
```

File: docs/en/tutorials/hook.md [hunk 2]
```diff
@@ -129,6 +130,14 @@ The four features mentioned above are described below.
   default_hooks = dict(checkpoint=dict(type='CheckpointHook', interval=1, save_best='accuracy', rule='less', published_keys=['meta', 'state_dict']))
   ```
 
+- Control the epoch number or iteration number at which checkpoint saving begins
+
+  If you want to set the number of epochs or iterations to control the start of saving weights, you can set the `save_begin` parameter, defaults to 1, which means saving checkpoints from the beginning of training. For example, if you train for a total of 10 epochs, and `save_begin` is set to 6, then the checkpoints for epochs 6, 7, 8, 9, and 10 will be saved. If `interval=2`, only save checkpoints for epochs 6, 8 and 10.
```
Comment (zhouzaida) [inline docs/en/tutorials/hook.md:135]: if `save_begin` is set to 5, and the interval is 2, which epochs should be saved?

Comment (zhouzaida) [reply docs/en/tutorials/hook.md:135]: 5, 7, 9
or
6, 8, 10

Comment (zhouzaida) [reply docs/en/tutorials/hook.md:135]: Suggest setting `save_begin` as 0 by default. Here is an example of saving checkpoints.

interval=2
- save_begin=0: 2, 4, 6, 8, 10   keep backward compatible.
- save_begin=1: 1, 3, 5, 7, 9
- save_begin=5: 5, 7, 9

Comment (KerwinKai) [reply docs/en/tutorials/hook.md:135]: oki, i will change my code

Comment (KerwinKai) [reply docs/en/tutorials/hook.md]: Already fix it.

File: mmengine/hooks/hook.py [hunk 3]
```diff
@@ -335,18 +335,20 @@ def _after_iter(self,
...
                 process.
             n (int): Whether current epoch can be evenly divided by n.
+            st(int): The parameter of save_begin, controlling the
+                epoch number at which checkpoint saving begins.
 
         Returns:
             bool: Whether current epoch can be evenly divided by n.
         """
-        return (runner.epoch + 1) % n == 0 if n > 0 else False
+        return (runner.epoch + 1 - st) % n == 0 if n > 0 else False
```
Comment (zhouzaida) [inline mmengine/hooks/hook.py:351]:
```suggestion
        dividend = runner.epoch + 1 - start
        return dividend % n == 0 if dividend >=0 and n > 0 else False
```

File: mmengine/hooks/hook.py [hunk 4]
```diff
@@ -363,19 +365,21 @@ def every_n_inner_iters(self, batch_idx: int, n: int) -> bool:
         """
         return (batch_idx + 1) % n == 0 if n > 0 else False
 
-    def every_n_train_iters(self, runner, n: int) -> bool:
+    def every_n_train_iters(self, runner, n: int, st: int = 0) -> bool:
```
Comment (zhouzaida) [inline mmengine/hooks/hook.py:368]: same as `every_n_epochs`

File: mmengine/hooks/checkpoint_hook.py [hunk 5]
```diff
@@ -325,10 +331,13 @@ def after_train_epoch(self, runner) -> None:
         if not self.by_epoch:
             return
 
+        if runner.epoch + 1 < self.save_begin:
+            return
```
Comment (zhouzaida) [inline mmengine/hooks/checkpoint_hook.py:335]: If the logic of `every_n_epochs` is updated as I proposed, this check can be deleted.

File: mmengine/hooks/checkpoint_hook.py [hunk 6]
```diff
@@ -244,6 +250,10 @@ def __init__(self,
         self.published_keys = published_keys
 
         self.last_ckpt = None
+        if save_begin < 0:
+            raise ValueError(
+                'save_begin should not less than 0, but got {save_begin}')
```
Comment (zhouzaida) [inline mmengine/hooks/checkpoint_hook.py:255]:
```suggestion
                'save_begin should not be less than 0, but got {save_begin}')
```

File: mmengine/hooks/checkpoint_hook.py (hunk 5 above)
Comment (zhouzaida) [reply mmengine/hooks/checkpoint_hook.py:339]: Hi @KerwinKai , this is the final comment. Thanks for your contribution.

**Review by zhouzaida — APPROVED** [review]

**Comment by zhouzaida** [comment]
Hi @KerwinKai , the unit tests failed. https://github.com/open-mmlab/mmengine/actions/runs/5655175431/job/15319783064?pr=1271

**Pull request merged**